datanode/dehistory/initialise.go

```diff
@@ -215,33 +224,43 @@ func GetMostRecentHistorySegmentFromPeer(ctx context.Context, ip string, datanod
 		return nil, fmt.Errorf("failed to get most recent history segment:%w", err)
 	}
 
-	return resp.GetSegment(), nil
+	return resp, nil
 }
 
 // TODO this needs some thought as to the best strategy to select the root segment to avoid spoofing.
-func SelectRootSegment(peerToSegment map[string]*v2.HistorySegment) *v2.HistorySegment {
-	segmentsList := make([]*v2.HistorySegment, 0, len(peerToSegment))
+func SelectMostRecentHistorySegmentResponse(peerToSegment map[string]*v2.GetMostRecentDeHistorySegmentResponse, swarmKey string) *PeerResponse {
+	segmentsList := make([]PeerResponse, 0, len(peerToSegment))
 
-	for _, segment := range peerToSegment {
-		segmentsList = append(segmentsList, segment)
+	for peer, segment := range peerToSegment {
+		segmentsList = append(segmentsList, PeerResponse{peer, segment})
 	}
 
 	// Sort history latest first
 	sort.Slice(segmentsList, func(i, j int) bool {
-		return segmentsList[i].ToHeight > segmentsList[j].ToHeight
+		return segmentsList[i].Response.Segment.ToHeight > segmentsList[j].Response.Segment.ToHeight
 	})
 
+	// Filter out segments without matching swarm key
+	var filteredSegments []PeerResponse
+	for _, segment := range segmentsList {
+		if segment.Response.SwarmKey == swarmKey {
+			filteredSegments = append(filteredSegments, segment)
+		}
+	}
+
 	// Filter out segments with toHeight < highest toHeight
-	var filteredSegments []*v2.HistorySegment
+	segmentsList = filteredSegments
+	filteredSegments = nil
 	for _, segment := range segmentsList {
-		if segment.ToHeight == segmentsList[0].ToHeight {
+		if segment.Response.Segment.ToHeight == segmentsList[0].Response.Segment.ToHeight {
 			filteredSegments = append(filteredSegments, segment)
 		}
 	}
 
 	// Select one segment from the list at random
 	if len(filteredSegments) > 0 {
-		return filteredSegments[rand.Intn(len(filteredSegments))]
+		segment := filteredSegments[rand.Intn(len(filteredSegments))]
```

Comment:
Why use a random one? is it possible to have different height here? No way to have a more consistent behaviour?

Reply:
Indeed, it probably will be random to an extent anyway depending on how and when each peer is contacted, but as per the comment on this method:

`// TODO this needs some thought as to the best strategy to select the root segment to avoid spoofing.`

might be that we could compare and pick the one where there is a majority agreement for example, and not necessarily the highest height (as currently) for example.  But that not's perfect either, so just opted to ensure it's totally random pick from the highest height set for now.  maybe we could have  some sort of consensus process whereby the validators decided amonst them what the are valid segments that way we could have a check on this, but maybe that is too much?